File: Codeforces/871d4/e/e.py
# After setting recursionlimit, we have a segfault.

from sys import setrecursionlimit

dirs = [(1, 0), (0, 1), (-1, 0), (0, -1)]

def solve():
    r = 1000; c = 1000
    arr = [[1000 for _ in range(1000)] for _ in range(1000)]
    # r, c = list(map(int, input().split(' ')))
    # arr = []

    # for _ in range(r):
    #     arr.append(list(map(int, input().split(' '))))

    seen = set()
    best = 0

    def dfs(v):
        if v in seen:
            return 0
        ans = arr[v[0]][v[1]]
        seen.add(v)
        for d in dirs:
            p, q = v[0] + d[0], v[1] + d[1]
            if 0 <= p < r and 0 <= q < c and arr[p][q] != 0:
                ans += dfs((p, q))

        return ans


    for i in range(r):
        for j in range(c):
            if not arr[i][j] == 0 and (i, j) not in seen:
                best = max(best, dfs((i, j)))

    print(best)


if __name__ == "__main__":
    # setrecursionlimit(1000000)
    t = int(input())
    # t = 1
    for _ in range(t):
        solve()

File: Codeforces/871d4/e/e.cpp
#include <iostream>
#include <vector>

#define MAX_N 1000
#define MAX_M 1000

using namespace std;

int arr[MAX_M][MAX_N];
int vis[MAX_M][MAX_N];
int n, m;

int dfs(int i, int j) {
	if (vis[i][j] || arr[i][j] == 0) return 0;
	int ret = arr[i][j];
	vis[i][j] = 1;
	if (i + 1 < n && arr[i + 1][j] != 0) ret += dfs(i + 1, j);
	if (i - 1 >= 0 && arr[i - 1][j] != 0) ret += dfs(i - 1, j);
	if (j + 1 < m && arr[i][j + 1] != 0) ret += dfs(i, j + 1);
	if (j - 1 >= 0 && arr[i][j - 1] != 0) ret += dfs(i, j - 1);
	return ret;
}

void solve() {
	cin >> n >> m;
	for (int i = 0; i < n; ++i)
	{
		for (int j = 0; j < m; ++j)
		{
			cin >> arr[i][j];
			vis[i][j] = 0;
		}
	}
	int best = 0;
	for (int i = 0; i < n; ++i)
		for (int j = 0; j < m; ++j)
			if (arr[i][j] != 0)
				best = max(best, dfs(i, j));
	cout << best << '\n';
}

int main() {
	int t;
	cin >> t;
	while (t--) {
		solve();
	}
	return 0;
}
